shared/src/main/scala/mlscript/MLParser.scala

```diff
@@ -304,7 +305,186 @@ class MLParser(origin: Origin, indent: Int = 0, recordLocations: Bool = true) {
     curHash = res + 1
     res
   }
-  
+
+  ///////////////////////////////////////////////////////
+  /// ADT types
+  ///////////////////////////////////////////////////////
+
+  // Param defined like 'a
+  def adtTyParam[p: P]: P[TypeName] = locate(P(("'" ~~ lowercase.!).map(param =>
+    TypeName("'" + param)
+  )))
+
+  // Params defined like ('a, 'b) or 'a
```

Comment:
Can we use the existing type parameter syntax, for uniformity? As in: `List[A]` instead of `'a List`.

Reply:
Well the problem with using the existing syntax is that `Int` and `A` are the same. So a constructor like `Tup of Int * A` doesn't know by construction that it uses a type variable in it's body. This information is needed to fill the `paramPos` field in `AdtInfo`.

For e.g. in `datatype Tup[A, B] = IntTup of Int * B`, the `paramPos` for `IntTup` will be `List(1)`.

```
final case class AdtInfo(alsName: TypeName, paramPos: Ls[Int])
```

Maybe I can add a separate step that traverses the term tree and finds all the type variables by checking that it's a single capitalized alphabetic character.

Reply:
On the other hand I can invent new syntax and simplify the whole thing with something like this

```
datatype Either[A, B] = Left[A] of A | Right[B] of B
datatype Tup[A, B] = IntTup[B] of Int * B | StrTup[A] of A * Str
```

This way the constructor has the information of what is parameterized.

Reply:
I don't understand why you think supporting `datatype Either[A, B]` is harder than supporting `('a, 'b) either`. The two syntaxes are isomorphic and contain the same amount of information. What's the problem?



> I can invent new syntax and simplify the whole thing with something like this
> 
> ```
> datatype Either[A, B] = Left[A] of A | Right[B] of B
> ```

I don't think we need that. Why would `Left` and `Right` need their own type parameter lists? That's not what happens in the current error messages prototype.

To be clear, we _do not want_ for `Left` and `Right` to introduce their own types. Remember we want ML-style typing of these ADTs, meaning the only type here is `Either[A, B]` and `Left`/`Right` are just constructors, to be used in expressions and patterns. OTOH, we still want to generate `Left`/`Right` classes but that's just for the purpose of code generation. Their ctors need to be shadowed with traditional the ML signature.

Reply:
So I think the generated code should look like:

```
class Either[A, B]
class Left[A, B](value: A) extends Either[A, B]
class Right[A, B](value: B) extends Either[A, B]
def Left[A, B] = Left : A -> Either[A, B]
def Right[A, B] = Right : B -> Either[A, B]
```
Notice the upcasts in the defs shadowing the case ctors.

Reply:
data declarations don't create new types only new constructor functions
The information is still captured in the `Either` typedef `AdtInfo` so that it can be used for codegen